File: Python/UTF-8_Validation.py
class Solution(object):
    def validUtf8(self, data):
        """
        :type data: List[int]
        :rtype: bool
        """
        if len(data) <= 0: return False
        mask = 1 << 7
        if len(data) == 1: return not data[0]&mask == mask
        count = 0
        for d in data:
            if count == 0:
                while d&mask == mask:
                    count += 1
                    d <<= 1
                if count == 1: return False
                count = max(count-1, 0)
            else:
                if d&mask == mask and (d<<1)&mask == 0:
                    count -= 1
                else: return False
        return count == 0

File: Go/UTF-8_Validation.go
func validUtf8(data []int) bool {
    if len(data) == 0 {
        return false
    }
    x,mask := 0,1 << 7
    count := 0
    for i := 0; i < len(data); i ++ {
        x = data[i]
        if count == 0 {
            for x&mask == mask {
                x <<= 1
                count ++
            }
            if count > 0 {
                count --
                if count == 0 {
                    return false
                }
            }
        } else {
            if x & mask == mask && (x<<1) & mask == 0 {
                count --;
            } else {
                return false
            }
        }
    }
    return count == 0
}
